### PR TITLE
Resolve: BUG: Pool Cycle not overridden by Mimi

### LINKED ISSUE #1327: BUG: Pool Cycle not overridden by Mimi
If over-ridden by Mimir, use that value instead.

https://thornode.thorchain.info/thorchain/mimir

`PoolCycle`

#### Extra notes:
Currently ASGDX uses `PoolCycle` from Midgard to count blocks for pending pools in [getBlocksLeftForPendingPool](https://github.com/thorchain/asgardex-electron/blob/86cc864a9addbf21ccc3473f920ef0030a917971/src/renderer/views/pools/Pools.utils.ts#L91)

### PR DESCRIPTION
**Pull request by thatStrangeGuyThorchain**
- `liveData`: introduced `chainOnError`
- introduced `usePoolCycle` hook
- refactored `PendingPools` to use `usePoolCycle`

https://user-images.githubusercontent.com/67793194/120818483-572a0700-c55b-11eb-9985-7fa29530973a.mov

closes #1327

### CODE REVIEW

**Review by veado — CHANGES_REQUESTED** [review]
As discussed in DM: Do a request to` thorchain/mimi` to get pool cycle. Use it if available, if not do another request to Midgard and use it from there. 

One way might be to create a `usePoolCycle` hook, which handles` mimir$` + `thorchainConstantsState$` to get the value and use it in the view. The hook can also provide a `reloadPoolCycle` function, which does a call for all needed reload functions to trigger  `mimir$` / `thorchainConstantsState$` updates.

File: src/renderer/hooks/usePoolCycle.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+          O.fromNullable(newPoolCycle),
+          O.fold(
+            (): LiveData<Error, number> => Rx.of(RD.failure(Error(`No POOLCYCLE at the mimir's response`))),
+            (poolCycle) => liveData.of(poolCycle)
+          )
+        )
+      ),
+      liveData.chainOnError(getPoolCycleFromMidgard),
+      liveData.mapLeft(() => ({
+        errorId: ErrorId.GET_MIMIR,
```
Comment (veado) [inline src/renderer/hooks/usePoolCycle.ts]: `ErrorId.GET_POOL_CYCLE` ?

File: src/renderer/hooks/usePoolCycle.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        msg: 'Unable to load pool cycle'
```
Comment (veado) [inline src/renderer/hooks/usePoolCycle.ts]: `Unable to load pool cycle from Midgard`

File: src/renderer/hooks/usePoolCycle.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  }, [reloadMimir, mimir$])
+
+  const getData = useCallback(() => {
+    FP.pipe(
+      getPoolCycleFromMimir(),
+      liveData.chain(({ 'mimir//POOLCYCLE': newPoolCycle }) =>
+        FP.pipe(
+          O.fromNullable(newPoolCycle),
+          O.fold(
+            (): LiveData<Error, number> => Rx.of(RD.failure(Error(`No POOLCYCLE at the mimir's response`))),
```
Comment (veado) [inline src/renderer/hooks/usePoolCycle.ts]: `Error('Unable to load pool cycle from Mimir')`

File: src/renderer/views/pools/PendingPools.tsx [hunk 4]
```diff
@@ -62,14 +60,20 @@ export const PendingPools: React.FC<PoolsComponentProps> = (): JSX.Element => {
...
 
+  const [poolCycle, reloadPpoolCycle] = usePoolCycle()
+
+  const oNewPoolCycle = useMemo(() => FP.pipe(poolCycle, RD.toOption), [poolCycle])
+
   const refreshHandler = useCallback(() => {
     reloadPendingPools()
     reloadFundsCap()
-  }, [reloadFundsCap, reloadPendingPools])
+    reloadPpoolCycle()
```
Comment (veado) [inline src/renderer/views/pools/PendingPools.tsx]: typo `reloadPoolCycle`

File: src/renderer/views/pools/PendingPools.tsx [hunk 5]
```diff
@@ -62,14 +60,20 @@ export const PendingPools: React.FC<PoolsComponentProps> = (): JSX.Element => {
...
+  }, [reloadFundsCap, reloadPendingPools, reloadPpoolCycle])
 
   const pendingCountdownHandler = useCallback(() => {
     reloadThorchainLastblock()
-  }, [reloadThorchainLastblock])
+    reloadPpoolCycle()
```
Comment (veado) [inline src/renderer/views/pools/PendingPools.tsx]: Same typo ^

File: src/renderer/hooks/usePoolCycle.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    () => FP.pipe(poolCycleState$, RxOp.shareReplay(1), RxOp.distinctUntilChanged(eqPoolCycle.equals)),
+    RD.initial
+  )
+
+  useEffect(() => {
+    getData()
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [])
+
+  return [data, getData]
```
Comment (veado) [inline src/renderer/hooks/usePoolCycle.ts]: Return a `object` (not an array)
`{data: poolCycle, getData: reloadPoolCycle}`

File: src/renderer/hooks/usePoolCycle.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            (poolCycle) => liveData.of(poolCycle)
+          )
+        )
+      ),
+      liveData.chainOnError(getPoolCycleFromMidgard),
+      liveData.mapLeft(() => ({
+        errorId: ErrorId.GET_MIMIR,
+        msg: 'Unable to load pool cycle'
+      }))
+    ).subscribe(setPoolCycleState)
```
Comment (veado) [inline src/renderer/hooks/usePoolCycle.ts]: Any way to unsubscribe it later in `useEffect` (L73)?

**Review by veado — CHANGES_REQUESTED** [review]
Question about `usePoolCycle`: 

Isn't there a more simple way just to create a `useObservableState` for `poolcycle` by subscribing to `thorchainConstantsState$` /  `mimir$` at once to get and compare all needed data from both streams? Similar to what we have in `useFundsCap` (not the same because of different data, but you might get the idea). And for reloading just 
```
reloadPoolCycle = () => {
 reloadThorchainConstants()
reloadMimir()
}
```

File: src/renderer/views/pools/PendingPools.tsx (hunk 5 above)
Comment (veado) [reply src/renderer/views/pools/PendingPools.tsx]: By `reloadPoolCycle()` in `pendingCountdownHandler` we will load a constant (pool cycle) every 5 seconds - we don't need it. It's ok to load this constant just once whenever an users enters `PendingPools`.

File: src/renderer/hooks/usePoolCycle.ts [hunk 8]
```diff
@@ -12,68 +12,68 @@ import { useMidgardContext } from '../contexts/MidgardContext'
...
   )
 
+  // Reload data on mount
   useEffect(() => {
     getData()
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [])
 
-  return [data, getData]
+  return useMemo(() => ({ data, getData }), [data, getData])
```
Comment (veado) [inline src/renderer/hooks/usePoolCycle.ts]: Why `useMemo`?

**Review by veado — CHANGES_REQUESTED** [review]
Pls check prev. comments (e.g. https://github.com/thorchain/asgardex-electron/pull/1504#discussion_r645661814, https://github.com/thorchain/asgardex-electron/pull/1504#discussion_r645634637).

File: src/renderer/hooks/usePoolCycle.ts (hunk 8 above)
Comment (thatStrangeGuyThorchain) [reply src/renderer/hooks/usePoolCycle.ts]: just to avoid possible re-creating of object

**Review by veado — APPROVED** [review]
:+1:

**Pull request merged**